ironfish/src/network/peers/addressManager.ts

```diff
@@ -2,24 +2,149 @@
+    )
+
+    if (dedupedAddresses.length) {
+      this.hostsStore.set('possiblePeers', [...this.possiblePeerAddresses, ...dedupedAddresses])
+    }
+  }
+
+  /**
+   * Returns a peer address for a disconnected peer by using current peers to
+   * filter out peer addresses. It attempts to find a previously-connected
+   * peer address that is not part of an active connection. If there are none,
+   * then it attempts to find a previously-known peer address.
+   */
+  getRandomDisconnectedPeerAddress(peers: Peer[]): PeerAddress | null {
+    if (
+      this.possiblePeerAddresses.length === 0 &&
+      this.priorConnectedPeerAddresses.length === 0
+    ) {
+      return null
+    }
+
+    const currentPeerAddresses = new Set(
+      //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      peers.filter((peer) => peer.state.identity !== null).map((peer) => peer.state.identity!),
+    )
+
+    const priorConnectedAddressSet = new Set([...this.priorConnectedPeerAddresses])
+
+    const disconnectedPriorAddresses = this.filterConnectedAddresses(
+      priorConnectedAddressSet,
+      currentPeerAddresses,
+    )
+    if (disconnectedPriorAddresses.length) {
+      return ArrayUtils.sampleOrThrow(disconnectedPriorAddresses)
+    } else {
+      const possibleAddressSet = new Set([...this.possiblePeerAddresses])
+      const disconnectedPossibleAddresses = this.filterConnectedAddresses(
+        possibleAddressSet,
+        currentPeerAddresses,
+      )
+      if (disconnectedPossibleAddresses.length) {
+        return ArrayUtils.sampleOrThrow(disconnectedPossibleAddresses)
+      } else {
+        return null
+      }
+    }
+  }
+
+  private filterConnectedAddresses(
+    addressSet: Set<Readonly<PeerAddress>>,
+    connectedPeerAdresses: Set<string>,
+  ): PeerAddress[] {
+    const disconnectedAddresses = [...addressSet].filter(
+      (address) => address.identity !== null && !connectedPeerAdresses.has(address.identity),
+    )
+
+    return disconnectedAddresses
+  }
+
+  /**
+   * Removes address associated with a peer from address stores
+   */
+  removePeerAddress(peer: Peer): void {
+    const filteredPossibles = this.possiblePeerAddresses.filter(
+      (possible) => possible.identity !== peer.state.identity,
+    )
+    const filteredPriorConnected = this.priorConnectedPeerAddresses.filter(
+      (prior) => prior.identity !== peer.state.identity,
+    )
+
+    this.hostsStore.set('possiblePeers', filteredPossibles)
+    this.hostsStore.set('priorPeers', filteredPriorConnected)
+  }
+
+  /**
+   * Persist all currently connected peers and unused peer addresses to disk
+   */
+  async save(peers: Peer[]): Promise<void> {
+    const inUsePeerAddresses = peers
+      .filter(
+        (peer) =>
+          peer.state.type === 'CONNECTED' &&
+          !peer.getConnectionRetry(ConnectionType.WebSocket, ConnectionDirection.Outbound)
+            ?.willNeverRetryConnecting,
+      )
```

Comment:
This is a TS-enforced assertion, but I think the `?.` operator isn't necessary here if the ConnectionDirection is Outbound.